WordPress/wordpress-develop
# Add `__experimentalFeatures` data for the block editor

This PR ports the first piece of landing the theme.json processing in WordPress core. It deals with merging the new settings (`__experimentalFeatures`) for all editors to access. The second piece would be to land the stylesheet engine.

## What this PR does

- Adds a new setting `__experimentalFeatures` to the block editor settings. This data comes from `theme.json` (core and theme). [See](https://developer.wordpress.org/block-editor/how-to-guides/themes/theme-json/) for the whole rationale about the `theme.json` file.
- Adds a `theme-i18n.json` file to WordPress core, to be used by core, the plugin, and wp-cli for translations.
- Adds the necessary code to make this work (namely the `WP_Theme_JSON` and `WP_Theme_JSON_Resolver` classes. These are used to pull settings from `theme.json` (core and theme). If you compare these with the ones that come with Gutenberg you'll note that this PR only includes the necessary bits for settings to work. Subsequent PRs will deal with adding styles, etc.

## How this can be extended

There are a few ways:

- For major changes (change how core and theme data is merged, changes to the shape of `theme.json` file, etc): because we're adding new data under `settings.__experimentalFeatures` this can be filtered by the existing block editor filters.
- To add more settings and its default values: two new filters are introduced.
  - `theme_json_allowed_schema`: to modify the schema used in sanitize the `theme.json` data.
  - `theme_json_core_data`: to set defaults for core data, including new settings added via the previous filter.

## Related work

Dependant PRs:

- Add `_wp_array_set` function and extract it from this PR https://github.com/WordPress/wordpress-develop/pull/1263 (to land first)
- Adapt Gutenberg for this changes https://github.com/WordPress/gutenberg/pull/32059

Follow-ups:

- Prepare a similar PR to add the styles engine.
- Port the changes to the merge algorithm introduced at https://github.com/WordPress/gutenberg/pull/31669 when the Gutenberg 10.7 packages are ported to core.

## How to test

A theme without `theme.json`:

- Activate the TwentyTwentyOne theme (doesn't have a `theme.json`).
- Go to the editor and make sure that there's a new key `__experimentalFeatures` available under `settings` in the `core/block-editor` store and that it contains the ALLOWED_SETTINGS ([see](https://github.com/WordPress/wordpress-develop/pull/1262/files#diff-e12c3008d747d1bec5be9927c5e7b1ced0a88641abe52c278d495da936714817R26))
- A good litmus test is that the color palette is shown properly in the UI controls. Verify it's the same defined by the theme.

A theme with `theme.json` support (but doesn't provide specific color palette):

- Install and activate the emptytheme.
- Go to the editor and make sure that there's a new key `__experimentalFeatures` available under `settings` in the `core/block-editor` store and that it contains the ALLOWED_SETTINGS ([see](https://github.com/WordPress/wordpress-develop/pull/1262/files#diff-e12c3008d747d1bec5be9927c5e7b1ced0a88641abe52c278d495da936714817R26))
- Check that the color palette shown is the one by core.

A theme with `theme.json` support (but does provide a color palette):

- Install and activate the TT1-blocks theme.
- Go to the editor and make sure that there's a new key `__experimentalFeatures` available under `settings` in the `core/block-editor` store and that it contains the ALLOWED_SETTINGS ([see](https://github.com/WordPress/wordpress-develop/pull/1262/files#diff-e12c3008d747d1bec5be9927c5e7b1ced0a88641abe52c278d495da936714817R26))
- Check that the color palette shown is the one by the theme.

With the Gutenberg plugin active:

- Use this branch https://github.com/WordPress/gutenberg/pull/32059
- Go to the editor and make sure that there's a new key `__experimentalFeatures` available under `settings` in the `core/block-editor` store and that it contains the allowed settings passed from the plugin (see `gutenberg_global_styles_theme_json_allowed_settings` method).
